src/Internal/FlatMapOperation.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+namespace Amp\Pipeline\Internal;
+
+use Amp\Pipeline\ConcurrentIterableIterator;
+use Amp\Pipeline\ConcurrentIterator;
+use function Amp\async;
+use function Amp\Future\await;
+
+/** @internal */
+final class FlatMapOperation implements IntermediateOperation
+{
+    public static function getStopMarker(): object
+    {
+        static $marker;
+
+        return $marker ??= new \stdClass;
+    }
+
+    private int $concurrency;
+
+    private bool $ordered;
+
+    private \Closure $flatMap;
+
+    public function __construct(int $concurrency, bool $ordered, \Closure $flatMap)
+    {
+        $this->concurrency = $concurrency;
+        $this->ordered = $ordered;
+        $this->flatMap = $flatMap;
+    }
+
+    public function __invoke(ConcurrentIterator $source): ConcurrentIterator
+    {
+        $destination = new QueueState;
+        $stop = self::getStopMarker();
+
+        if ($this->concurrency === 1) {
+            return new ConcurrentIterableIterator((function () use ($source, $stop): iterable {
+                foreach ($source as $position => $value) {
+                    $iterable = ($this->flatMap)($value, $position);
+                    foreach ($iterable as $item) {
+                        if ($item === $stop) {
+                            return;
+                        }
+
+                        yield $item;
+                    }
+                }
+            })());
```

Comment:
I've changed this to use `ConcurrentIterableIterator`

Reply:
Does this introduce any difference between concurrency > 1? Should each fiber below use `ConcurrentIterableIterator` too?

Reply:
It makes it lazy, too, yes. We could use a slightly modified version for the concurrent version, yes.